File: apps/go-backend/deployment_lane_registry.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"strings"
)

const deploymentLaneRegistrySchemaVersion = "runtime-deployment-lane-registry-v1.37"

var runtimeSuccessorCanonicalTupleV117 = func() canonicalCompatibilityTuple {
	var tuple canonicalCompatibilityTuple
	_ = json.Unmarshal([]byte(runtimeSuccessorSemanticTupleV117), &tuple)
	return tuple
}()

type goDeploymentLaneProfile struct {
	ProviderID                       string                                `json:"providerId"`
	LanguageID                       string                                `json:"languageId"`
	LanguageVersion                  string                                `json:"languageVersion"`
	RuntimeID                        string                                `json:"runtimeId"`
	RuntimeVersion                   string                                `json:"runtimeVersion"`
	ToolchainID                      string                                `json:"toolchainId"`
	ToolchainVersion                 string                                `json:"toolchainVersion"`
	AdapterID                        string                                `json:"adapterId"`
	AdapterVersion                   string                                `json:"adapterVersion"`
	PolicyID                         string                                `json:"policyId"`
	PolicyVersion                    string                                `json:"policyVersion"`
	CorpusID                         string                                `json:"corpusId"`
	CorpusVersion                    string                                `json:"corpusVersion"`
	ArtifactKind                     string                                `json:"artifactKind"`
	ArtifactIDPrefix                 string                                `json:"artifactIdPrefix"`
	ImplementationID                 string                                `json:"implementationId"`
	BuildID                          string                                `json:"buildId"`
	SemanticTupleID                  string                                `json:"semanticTupleId"`
	SemanticTuple                    canonicalCompatibilityTuple           `json:"semanticTuple"`
	SuccessorRuntimeIdentityTemplate *runtimeSuccessorIdentityTemplateV117 `json:"successorRuntimeIdentityTemplate,omitempty"`
}

type goDeploymentLaneRegistry struct {
	SchemaVersion string                    `json:"schemaVersion"`
	RegistryID    string                    `json:"registryId"`
	Lanes         []goDeploymentLaneProfile `json:"lanes"`
}

func loadDeploymentLaneRegistry(path string) (*goDeploymentLaneRegistry, error) {
	if strings.TrimSpace(path) == "" {
		return nil, errors.New("deployment lane registry path is required")
	}
	bytes, err := os.ReadFile(path)
	if err != nil {
		return nil, errors.New("deployment lane registry could not be loaded")
	}
	var registry goDeploymentLaneRegistry
	if err := decodeStrictJSON(bytes, &registry); err != nil || registry.SchemaVersion != deploymentLaneRegistrySchemaVersion || strings.TrimSpace(registry.RegistryID) == "" || len(registry.Lanes) == 0 {
		return nil, errors.New("deployment lane registry is invalid")
	}
	keys := map[string]bool{}
	for _, lane := range registry.Lanes {
		values := []string{lane.ProviderID, lane.LanguageID, lane.LanguageVersion, lane.RuntimeID, lane.RuntimeVersion, lane.ToolchainID, lane.ToolchainVersion, lane.AdapterID, lane.AdapterVersion, lane.PolicyID, lane.PolicyVersion, lane.CorpusID, lane.CorpusVersion, lane.ArtifactIDPrefix, lane.ImplementationID, lane.BuildID, lane.SemanticTupleID, lane.SemanticTuple.Rules, lane.SemanticTuple.Engine, lane.SemanticTuple.RuntimeABI, lane.SemanticTuple.Chronicle, lane.SemanticTuple.ArenaCatalog, lane.SemanticTuple.SetPolicy}
		for _, value := range values {
			if strings.TrimSpace(value) == "" {
				return nil, errors.New("deployment lane registry profile is incomplete")
			}
		}
		if lane.ArtifactKind != "source" && lane.ArtifactKind != "compiled" {
			return nil, errors.New("deployment lane registry artifact kind is invalid")
		}
		if !validDeploymentLaneSemanticTuple(lane.SemanticTupleID, lane.SemanticTuple) {
			return nil, errors.New("deployment lane registry semantic tuple identity is invalid")
		}
		if lane.SemanticTupleID == runtimeSuccessorSemanticTupleIDV117 {
			if lane.SuccessorRuntimeIdentityTemplate == nil || !normalizeRuntimeSuccessorIdentityTemplateV117(lane.SuccessorRuntimeIdentityTemplate) ||
				!successorIdentityTemplateMatchesLaneProfileV117(lane.SuccessorRuntimeIdentityTemplate, lane) {
				return nil, errors.New("deployment lane registry successor identity template is invalid")
			}
		} else if lane.SuccessorRuntimeIdentityTemplate != nil {
			return nil, errors.New("deployment lane registry successor identity template is mixed-version")
		}
		key := strings.Join([]string{lane.LanguageID, lane.LanguageVersion, lane.AdapterID, lane.AdapterVersion}, "\x00")
		if keys[key] {
			return nil, errors.New("deployment lane registry profile is ambiguous")
		}
		keys[key] = true
	}
	return &registry, nil
}

func (registry *goDeploymentLaneRegistry) successorIdentityTemplateForRevision(strategy runtimeServiceStrategyRevision) (*runtimeSuccessorIdentityTemplateV117, bool) {
	if registry == nil {
		return nil, false
	}
	language := mapValue(strategy.Runtime, "language")
	adapter := mapValue(strategy.Runtime, "adapter")
	languageID, languageVersion := stringValue(language, "id"), stringValue(language, "version")
	adapterID, adapterVersion := stringValue(adapter, "id"), stringValue(adapter, "version")
	var resolved *runtimeSuccessorIdentityTemplateV117
	for index := range registry.Lanes {
		profile := &registry.Lanes[index]
		if profile.LanguageID == languageID && profile.LanguageVersion == languageVersion && profile.AdapterID == adapterID && profile.AdapterVersion == adapterVersion {
			if resolved != nil || profile.SuccessorRuntimeIdentityTemplate == nil ||
				!successorIdentityTemplateMatchesLaneProfileV117(profile.SuccessorRuntimeIdentityTemplate, *profile) {
				return nil, false
			}
			resolved = profile.SuccessorRuntimeIdentityTemplate
		}
	}
	if resolved == nil {
		return nil, false
	}
	return cloneRuntimeSuccessorIdentityTemplateV117(resolved), true
}

func loadProductionDeploymentLaneRegistryFromEnvironment() (*goDeploymentLaneRegistry, error) {
	return loadDeploymentLaneRegistry(strings.TrimSpace(os.Getenv("COWARDS_RUNTIME_DEPLOYMENT_LANE_REGISTRY")))
}

func (registry *goDeploymentLaneRegistry) matchesAuthority(authority *verifiedRuntimeEvidenceAuthority) bool {
	if registry == nil || authority == nil || len(registry.Lanes) == 0 {
		return false
	}
	for _, profile := range registry.Lanes {
		if !validDeploymentLaneSemanticTuple(profile.SemanticTupleID, profile.SemanticTuple) ||
			!validDeploymentLaneSemanticTuple(authority.CompatibilityTuple.TupleID, authority.CompatibilityTuple.Tuple) ||
			profile.SemanticTupleID != authority.CompatibilityTuple.TupleID || profile.SemanticTuple != authority.CompatibilityTuple.Tuple {
			return false
		}
	}
	return true
}

func (registry *goDeploymentLaneRegistry) resolveRevision(id string, sourceHash string, sourceBytes int, runtime map[string]any, engine map[string]any, validation map[string]any, metadata map[string]any, tuple registeredCompatibilityTuple) (*goExecutableLaneIdentity, bool) {
	if registry == nil || id == "" || sourceHash == "" || sourceBytes <= 0 {
		return nil, false
	}
	language := mapValue(runtime, "language")
	adapter := mapValue(runtime, "adapter")
	languageID, languageVersion := stringValue(language, "id"), stringValue(language, "version")
	adapterID, adapterVersion := stringValue(adapter, "id"), stringValue(adapter, "version")
	var profile *goDeploymentLaneProfile
	for index := range registry.Lanes {
		candidate := &registry.Lanes[index]
		if candidate.LanguageID == languageID && candidate.LanguageVersion == languageVersion && candidate.AdapterID == adapterID && candidate.AdapterVersion == adapterVersion {
			if profile != nil {
				return nil, false
			}
			profile = candidate
		}
	}
	if profile == nil || !validDeploymentLaneSemanticTuple(profile.SemanticTupleID, profile.SemanticTuple) || !validDeploymentLaneSemanticTuple(tuple.TupleID, tuple.Tuple) || profile.SemanticTupleID != tuple.TupleID || profile.SemanticTuple != tuple.Tuple || stringValue(runtime, "abiVersion") != tuple.Tuple.RuntimeABI || stringValue(engine, "spec") != tuple.Tuple.Rules || stringValue(engine, "engine") != tuple.Tuple.Engine {
		return nil, false
	}
	if profile.SemanticTupleID == runtimeSuccessorSemanticTupleIDV117 {
		if !validSuccessorStrategyRevisionV117(sourceHash, sourceBytes, runtime, engine, validation, metadata) ||
			profile.SuccessorRuntimeIdentityTemplate == nil ||
			!successorIdentityTemplateMatchesLaneProfileV117(profile.SuccessorRuntimeIdentityTemplate, *profile) {
			return nil, false
		}
	}
	provider := mapValue(metadata, "providerValidation")
	if stringValue(provider, "providerId") != profile.ProviderID {
		return nil, false
	}
	artifact := mapValue(metadata, "sourceArtifact")
	providerValid := sourceArtifactProviderValidationMatchesABI(metadata, sourceHash, sourceBytes, profile.ProviderID, profile.LanguageID, tuple.Tuple.RuntimeABI)
	if profile.ArtifactKind == "compiled" {
		artifact = mapValue(metadata, "compiledArtifact")
		providerValid = rustProviderValidationMatches(metadata, sourceHash, sourceBytes, profile.LanguageID)
	}
	if !providerValid || !isLowerSHA256(stringValue(artifact, "hash")) {
		return nil, false
	}
	toolchain := mapValue(artifact, "toolchain")
	if profile.ArtifactKind == "source" {
		if profile.SemanticTupleID == runtimeSuccessorSemanticTupleIDV117 {
			if stringValue(toolchain, "language") != profile.LanguageID || stringValue(toolchain, "runtime") != profile.ToolchainID || stringValue(toolchain, "runtimeVersion") != profile.ToolchainVersion {
				return nil, false
			}
		} else if stringValue(toolchain, "language") != profile.ToolchainID || stringValue(toolchain, "runtime") != profile.RuntimeID || stringValue(toolchain, "runtimeVersion") != profile.ToolchainVersion || profile.RuntimeVersion != profile.ToolchainVersion {
			return nil, false
		}
	} else if stringValue(toolchain, "compiler") != profile.ToolchainID || stringValue(toolchain, "compilerVersion") != profile.ToolchainVersion {
		return nil, false
	}
	identity := &goExecutableLaneIdentity{
		ProviderID: profile.ProviderID, LanguageID: profile.LanguageID,
		RuntimeID: profile.RuntimeID, RuntimeVersion: profile.RuntimeVersion,
		ToolchainID: profile.ToolchainID, ToolchainVersion: profile.ToolchainVersion,
		AdapterID: profile.AdapterID, AdapterVersion: profile.AdapterVersion,
		PolicyID: profile.PolicyID, PolicyVersion: profile.PolicyVersion,
		CorpusID: profile.CorpusID, CorpusVersion: profile.CorpusVersion,
		ArtifactID: profile.ArtifactIDPrefix + id, ArtifactSHA256: stringValue(artifact, "hash"),
		ImplementationID: profile.ImplementationID, BuildID: profile.BuildID,
		SemanticTupleID: profile.SemanticTupleID, SemanticTuple: profile.SemanticTuple,
	}
	return identity, true
}

func successorIdentityTemplateMatchesLaneProfileV117(template *runtimeSuccessorIdentityTemplateV117, profile goDeploymentLaneProfile) bool {
	if !validNormalizedRuntimeSuccessorIdentityTemplateV117(template) || template.ExactPins[1][1] != profile.RuntimeVersion {
		return false
	}
	bindings := runtimeIdentityBindingMapV117(template.Bindings)
	laneProfileHash, ok := hashSuccessorRuntimeLaneProfileV117(profile)
	return ok && template.LaneProfileSHA256 == laneProfileHash &&
		bindings["containmentPolicy"].PublicID == profile.PolicyID &&
		bindings["conformanceCorpus"].PublicID == profile.CorpusID
}

func hashSuccessorRuntimeLaneProfileV117(profile goDeploymentLaneProfile) (string, bool) {
	values := map[string]any{
		"providerId": profile.ProviderID, "languageId": profile.LanguageID, "languageVersion": profile.LanguageVersion,
		"runtimeId": profile.RuntimeID, "runtimeVersion": profile.RuntimeVersion,
		"toolchainId": profile.ToolchainID, "toolchainVersion": profile.ToolchainVersion,
		"adapterId": profile.AdapterID, "adapterVersion": profile.AdapterVersion,
		"policyId": profile.PolicyID, "policyVersion": profile.PolicyVersion,
		"corpusId": profile.CorpusID, "corpusVersion": profile.CorpusVersion,
		"artifactKind": profile.ArtifactKind, "artifactIdPrefix": profile.ArtifactIDPrefix,
		"implementationId": profile.ImplementationID, "buildId": profile.BuildID,
		"semanticTupleId": profile.SemanticTupleID, "semanticTuple": profile.SemanticTuple,
	}
	if len(runtimeSuccessorLaneProfileFieldsV117) != len(values) {
		return "", false
	}
	exact := make(map[string]any, len(values))
	for _, field := range runtimeSuccessorLaneProfileFieldsV117 {
		value, exists := values[field]
		if !exists {
			return "", false
		}
		exact[field] = value
	}
	encoded, err := runtimeInvocationV117CanonicalValue(exact)
	if err != nil {
		return "", false
	}
	hash := sha256.New()
	_, _ = hash.Write([]byte(runtimeSuccessorLaneProfileDomainV117))
	_, _ = hash.Write([]byte{0})
	_, _ = hash.Write(encoded)
	return "sha256:" + hex.EncodeToString(hash.Sum(nil)), true
}

func validSuccessorStrategyRevisionV117(sourceHash string, sourceBytes int, runtime map[string]any, engine map[string]any, validation map[string]any, metadata map[string]any) bool {
	if sourceHash == "" || sourceBytes <= 0 ||
		!runtimeInvocationV117ExactKeys(runtime, "abiVersion", "language", "adapter", "package", "requiredCapabilities", "limits") ||
		stringValue(runtime, "abiVersion") != strategyRuntimeABIVersionV117 ||
		!runtimeInvocationV117ExactKeys(mapValue(runtime, "language"), "id", "version") ||
		!runtimeInvocationV117ExactKeys(mapValue(runtime, "adapter"), "id", "version") ||
		!runtimeInvocationV117ExactKeys(engine, "spec", "engine") {
		return false
	}
	pkg := mapValue(runtime, "package")
	if !runtimeInvocationV117ExactKeys(pkg, "entrypoint", "mode") || stringValue(pkg, "entrypoint") != "default" || stringValue(pkg, "mode") != "none" || !emptyRuntimeCapabilities(runtime["requiredCapabilities"]) {
		return false
	}
	limits := mapValue(runtime, "limits")
	if !validSuccessorRuntimeLimitsV117(limits) || !validSuccessorStrategyValidationV117(validation, sourceHash, sourceBytes, runtime, engine) {
		return false
	}
	providerValidation := mapValue(metadata, "providerValidation")
	artifact := mapValue(metadata, "sourceArtifact")
	return stringValue(providerValidation, "contractVersion") == "runtime-provider-validation-v1.17" &&
		stringValue(providerValidation, "sourceHash") == sourceHash && intValue(providerValidation, "sourceBytes") == sourceBytes &&
		stringValue(providerValidation, "artifactHash") == stringValue(artifact, "hash") &&
		intValue(providerValidation, "artifactBytes") == intValue(artifact, "bytes") &&
		stringValue(artifact, "abiVersion") == strategyRuntimeABIVersionV117 &&
		stringValue(engine, "spec") == runtimeSuccessorCanonicalTupleV117.Rules &&
		stringValue(engine, "engine") == runtimeSuccessorCanonicalTupleV117.Engine
}

func validSuccessorStrategyValidationV117(validation map[string]any, sourceHash string, sourceBytes int, runtime map[string]any, engine map[string]any) bool {
	if !runtimeInvocationV117ExactKeys(validation, "valid", "errors", "warnings", "sourceBytes", "forbiddenPatterns", "sourceHash", "runtimeVersion", "engineCompatibility") ||
		stringValue(validation, "sourceHash") != sourceHash || intValue(validation, "sourceBytes") != sourceBytes ||
		stringValue(validation, "runtimeVersion") != stringValue(mapValue(runtime, "adapter"), "version") {
		return false
	}
	validationEngine := mapValue(validation, "engineCompatibility")
	if !runtimeInvocationV117ExactKeys(validationEngine, "spec", "engine") ||
		stringValue(validationEngine, "spec") != stringValue(engine, "spec") ||
		stringValue(validationEngine, "engine") != stringValue(engine, "engine") {
		return false
	}
	errors, errorsOK := validation["errors"].([]any)
	warnings, warningsOK := validation["warnings"].([]any)
	patterns, patternsOK := validation["forbiddenPatterns"].([]any)
	valid, validOK := validation["valid"].(bool)
	if !validOK || !errorsOK || !warningsOK || !patternsOK {
		return false
	}
	for _, pattern := range patterns {
		if _, ok := pattern.(string); !ok {
			return false
		}
	}
	for _, validationError := range errors {
		if !validSuccessorValidationIssueV117(validationError, "error") {
			return false
		}
	}
	for _, warning := range warnings {
		if !validSuccessorValidationIssueV117(warning, "warning") {
			return false
		}
	}
	if valid != (len(errors) == 0) || !valid {
		return false
	}
	return true
}

func validSuccessorValidationIssueV117(value any, severity string) bool {
	issue, ok := value.(map[string]any)
	if !ok || (severity != "error" && severity != "warning") || len(issue) < 3 || len(issue) > 9 ||
		stringValue(issue, "severity") != severity ||
		!runtimeSuccessorStrategyValidationCodeKnownV117(stringValue(issue, "code")) ||
		stringValue(issue, "message") == "" {
		return false
	}
	for key, candidate := range issue {
		switch key {
		case "code":
			if code, ok := candidate.(string); !ok || !runtimeSuccessorStrategyValidationCodeKnownV117(code) {
				return false
			}
		case "severity":
			if issueSeverity, ok := candidate.(string); !ok || issueSeverity != severity {
				return false
			}
		case "message", "pattern", "constraint", "remediation", "reference":
			if text, ok := candidate.(string); !ok || text == "" {
				return false
			}
		case "line":
			if integer, ok := runtimeInvocationV117Integer(candidate); !ok || integer <= 0 {
				return false
			}
		case "column":
			if integer, ok := runtimeInvocationV117Integer(candidate); !ok || integer < 0 {
				return false
			}
		default:
			return false
		}
	}
	return true
}

func validSuccessorRuntimeLimitsV117(limits map[string]any) bool {
	return runtimeInvocationV117ExactKeys(limits,
		"timeoutMs", "stdoutBytes", "stderrBytes", "sourceBytes", "strategyMemoryBytes", "soldierMemoryBytes", "objectivePayloadBytes",
		"environment", "filesystem", "network", "shell", "packagePolicy",
	) && intValue(limits, "timeoutMs") == 1000 && intValue(limits, "stdoutBytes") == 262144 &&
		intValue(limits, "stderrBytes") == 65536 && intValue(limits, "sourceBytes") == 65536 &&
		intValue(limits, "strategyMemoryBytes") == 32768 && intValue(limits, "soldierMemoryBytes") == 2048 &&
		intValue(limits, "objectivePayloadBytes") == 1024 && stringValue(limits, "environment") == "empty" &&
		stringValue(limits, "filesystem") == "none" && stringValue(limits, "network") == "disabled" &&
		stringValue(limits, "shell") == "disabled" && stringValue(limits, "packagePolicy") == "none"
}

func emptyRuntimeCapabilities(value any) bool {
	switch typed := value.(type) {
	case []any:
		return len(typed) == 0
	case []string:
		return len(typed) == 0
	default:
		return false
	}
}

func validDeploymentLaneSemanticTuple(tupleID string, tuple canonicalCompatibilityTuple) bool {
	return validKnownVersionedCompatibilityTuple(tupleID, tuple)
}

File: apps/runtime-service/src/deployment-lane-registry.ts
import { Buffer } from "node:buffer"
import { createHash } from "node:crypto"
import { readFileSync } from "node:fs"
import {
  CANDIDATE_RUNTIME_V117_SEMANTIC_TUPLE_ID,
  StrategyRevisionV117Schema,
  hashSuccessorRuntimeLaneProfileV117,
  resolveCandidateRuntimeV117SemanticTuple,
  resolveCanonicalCompatibilityTuple,
  type CanonicalCompatibilityTuple,
  type ExecutableLaneIdentity,
  type StrategyRevision,
  type StrategyRevisionV117,
} from "@cowards/spec"
import { RuntimeServiceConfigError } from "./runtime-config.js"
import {
  parseSuccessorRuntimeIdentityTemplateV117,
  type SuccessorRuntimeIdentityTemplateV117,
} from "./successor-runtime-identity.js"

export const DEPLOYMENT_LANE_REGISTRY_SCHEMA_VERSION =
  "runtime-deployment-lane-registry-v1.37"

const SHA256_PATTERN = /^[0-9a-f]{64}$/u

const PROFILE_STRING_FIELDS = [
  "providerId",
  "languageId",
  "languageVersion",
  "runtimeId",
  "runtimeVersion",
  "toolchainId",
  "toolchainVersion",
  "adapterId",
  "adapterVersion",
  "policyId",
  "policyVersion",
  "corpusId",
  "corpusVersion",
  "artifactIdPrefix",
  "implementationId",
  "buildId",
  "semanticTupleId",
] as const

export interface DeploymentLaneProfile {
  providerId: string
  languageId: string
  languageVersion: string
  runtimeId: string
  runtimeVersion: string
  toolchainId: string
  toolchainVersion: string
  adapterId: string
  adapterVersion: string
  policyId: string
  policyVersion: string
  corpusId: string
  corpusVersion: string
  artifactKind: "source" | "compiled"
  artifactIdPrefix: string
  implementationId: string
  buildId: string
  semanticTupleId: string
  semanticTuple: CanonicalCompatibilityTuple
  successorRuntimeIdentityTemplate?: SuccessorRuntimeIdentityTemplateV117
}

export interface DeploymentLaneRegistry {
  schemaVersion: typeof DEPLOYMENT_LANE_REGISTRY_SCHEMA_VERSION
  registryId: string
  lanes: readonly Readonly<DeploymentLaneProfile>[]
}

const isRecord = (value: unknown): value is Record<string, unknown> =>
  value !== null && typeof value === "object" && !Array.isArray(value)

const exactKeys = (
  value: Record<string, unknown>,
  keys: readonly string[],
  label: string,
): void => {
  const actual = Object.keys(value).sort()
  const expected = [...keys].sort()
  if (
    actual.length !== expected.length ||
    actual.some((key, index) => key !== expected[index])
  ) {
    throw new RuntimeServiceConfigError(`${label} has invalid fields.`)
  }
}

const parseProfile = (value: unknown, index: number): DeploymentLaneProfile => {
  if (!isRecord(value)) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} is invalid.`,
    )
  }
  exactKeys(
    value,
    [
      ...PROFILE_STRING_FIELDS,
      "artifactKind",
      "semanticTuple",
      ...(Object.hasOwn(value, "successorRuntimeIdentityTemplate")
        ? (["successorRuntimeIdentityTemplate"] as const)
        : []),
    ],
    `Deployment lane registry profile ${index}`,
  )
  for (const field of PROFILE_STRING_FIELDS) {
    if (typeof value[field] !== "string" || value[field].trim().length === 0) {
      throw new RuntimeServiceConfigError(
        `Deployment lane registry profile ${index} ${field} is invalid.`,
      )
    }
  }
  if (value.artifactKind !== "source" && value.artifactKind !== "compiled") {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} artifactKind is invalid.`,
    )
  }
  const selector = {
    tupleId: value.semanticTupleId as string,
    tuple: value.semanticTuple,
  }
  const current = resolveCanonicalCompatibilityTuple(selector)
  const candidate = resolveCandidateRuntimeV117SemanticTuple(selector)
  const resolved = current ?? candidate
  if (!resolved) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} semantic tuple is invalid.`,
    )
  }
  let successorRuntimeIdentityTemplate:
    | DeploymentLaneProfile["successorRuntimeIdentityTemplate"]
    | undefined
  if (Object.hasOwn(value, "successorRuntimeIdentityTemplate")) {
    try {
      successorRuntimeIdentityTemplate =
        parseSuccessorRuntimeIdentityTemplateV117(
          value.successorRuntimeIdentityTemplate as SuccessorRuntimeIdentityTemplateV117,
        )
    } catch {
      throw new RuntimeServiceConfigError(
        `Deployment lane registry profile ${index} successor identity is invalid.`,
      )
    }
  }
  const isSuccessor =
    resolved.tupleId === CANDIDATE_RUNTIME_V117_SEMANTIC_TUPLE_ID
  if (
    (isSuccessor && successorRuntimeIdentityTemplate === undefined) ||
    (!isSuccessor && successorRuntimeIdentityTemplate !== undefined)
  ) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} successor identity does not match its semantic tuple.`,
    )
  }
  const semanticTupleBinding = successorRuntimeIdentityTemplate?.bindings.find(
    ({ domain }) => domain === "semanticTuple",
  )
  const containmentPolicyBinding =
    successorRuntimeIdentityTemplate?.bindings.find(
      ({ domain }) => domain === "containmentPolicy",
    )
  const conformanceCorpusBinding =
    successorRuntimeIdentityTemplate?.bindings.find(
      ({ domain }) => domain === "conformanceCorpus",
    )
  if (
    isSuccessor &&
    (semanticTupleBinding?.publicId !== resolved.tupleId ||
      semanticTupleBinding.sha256 !== resolved.sha256)
  ) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} successor identity does not bind its semantic tuple.`,
    )
  }
  const successorPins =
    successorRuntimeIdentityTemplate === undefined
      ? undefined
      : new Map(successorRuntimeIdentityTemplate.exactPins)
  if (
    isSuccessor &&
    successorPins?.get("reportedVersion") !== value.runtimeVersion
  ) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} successor identity does not bind its runtime version.`,
    )
  }
  if (
    isSuccessor &&
    (containmentPolicyBinding?.publicId !== value.policyId ||
      conformanceCorpusBinding?.publicId !== value.corpusId)
  ) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} successor identity does not bind its policy and corpus.`,
    )
  }
  if (
    isSuccessor &&
    successorRuntimeIdentityTemplate?.laneProfileSha256 !==
      hashSuccessorRuntimeLaneProfileV117({
        providerId: value.providerId as string,
        languageId: value.languageId as string,
        languageVersion: value.languageVersion as string,
        runtimeId: value.runtimeId as string,
        runtimeVersion: value.runtimeVersion as string,
        toolchainId: value.toolchainId as string,
        toolchainVersion: value.toolchainVersion as string,
        adapterId: value.adapterId as string,
        adapterVersion: value.adapterVersion as string,
        policyId: value.policyId as string,
        policyVersion: value.policyVersion as string,
        corpusId: value.corpusId as string,
        corpusVersion: value.corpusVersion as string,
        artifactKind: value.artifactKind,
        artifactIdPrefix: value.artifactIdPrefix as string,
        implementationId: value.implementationId as string,
        buildId: value.buildId as string,
        semanticTupleId: resolved.tupleId,
        semanticTuple: { ...resolved.tuple },
      })
  ) {
    throw new RuntimeServiceConfigError(
      `Deployment lane registry profile ${index} successor identity does not bind its exact deployment profile.`,
    )
  }
  return Object.freeze({
    ...Object.fromEntries(
      PROFILE_STRING_FIELDS.map((field) => [field, value[field]]),
    ),
    artifactKind: value.artifactKind,
    semanticTuple: Object.freeze({ ...resolved.tuple }),
    ...(successorRuntimeIdentityTemplate === undefined
      ? {}
      : { successorRuntimeIdentityTemplate }),
  }) as unknown as DeploymentLaneProfile
}

export const parseDeploymentLaneRegistry = (
  value: unknown,
): Readonly<DeploymentLaneRegistry> => {
  if (!isRecord(value)) {
    throw new RuntimeServiceConfigError("Deployment lane registry is invalid.")
  }
  exactKeys(
    value,
    ["schemaVersion", "registryId", "lanes"],
    "Deployment lane registry",
  )
  if (
    value.schemaVersion !== DEPLOYMENT_LANE_REGISTRY_SCHEMA_VERSION ||
    typeof value.registryId !== "string" ||
    value.registryId.trim().length === 0 ||
    !Array.isArray(value.lanes) ||
    value.lanes.length === 0
  ) {
    throw new RuntimeServiceConfigError("Deployment lane registry is invalid.")
  }
  const lanes = value.lanes.map(parseProfile)
  const keys = new Set<string>()
  for (const lane of lanes) {
    const key = [
      lane.languageId,
      lane.languageVersion,
      lane.adapterId,
      lane.adapterVersion,
    ].join("\0")
    if (keys.has(key)) {
      throw new RuntimeServiceConfigError(
        "Deployment lane registry contains an ambiguous runtime profile.",
      )
    }
    keys.add(key)
  }
  return Object.freeze({
    schemaVersion: DEPLOYMENT_LANE_REGISTRY_SCHEMA_VERSION,
    registryId: value.registryId,
    lanes: Object.freeze(lanes),
  })
}

export const loadDeploymentLaneRegistry = (
  path: string,
): Readonly<DeploymentLaneRegistry> => {
  if (path.trim().length === 0) {
    throw new RuntimeServiceConfigError(
      "Runtime service requires COWARDS_RUNTIME_DEPLOYMENT_LANE_REGISTRY.",
    )
  }
  try {
    return parseDeploymentLaneRegistry(
      JSON.parse(readFileSync(path, "utf8")) as unknown,
    )
  } catch (error) {
    if (error instanceof RuntimeServiceConfigError) throw error
    throw new RuntimeServiceConfigError(
      "Runtime service deployment lane registry could not be loaded.",
    )
  }
}

const artifactBytesAndHash = (
  revision: StrategyRevision | StrategyRevisionV117,
  profile: DeploymentLaneProfile,
): { hash: string } | undefined => {
  const sourceArtifact = revision.metadata.sourceArtifact
  const compiledArtifact = revision.metadata.compiledArtifact
  const artifact =
    profile.artifactKind === "source" ? sourceArtifact : compiledArtifact
  if (
    !artifact?.bytesBase64 ||
    !SHA256_PATTERN.test(artifact.hash) ||
    artifact.sourceHash !== revision.sourceHash ||
    artifact.abiVersion !== profile.semanticTuple.runtimeAbi
  ) {
    return undefined
  }
  const bytes = Buffer.from(artifact.bytesBase64, "base64")
  if (
    bytes.byteLength !== artifact.bytes ||
    createHash("sha256").update(bytes).digest("hex") !== artifact.hash
  ) {
    return undefined
  }
  if (profile.artifactKind === "source") {
    const successor =
      profile.semanticTupleId === CANDIDATE_RUNTIME_V117_SEMANTIC_TUPLE_ID
    if (
      !sourceArtifact ||
      (successor
        ? sourceArtifact.toolchain.language !== profile.languageId ||
          sourceArtifact.toolchain.runtime !== profile.toolchainId ||
          sourceArtifact.toolchain.runtimeVersion !== profile.toolchainVersion
        : sourceArtifact.toolchain.language !== profile.toolchainId ||
          sourceArtifact.toolchain.runtime !== profile.runtimeId ||
          sourceArtifact.toolchain.runtimeVersion !==
            profile.toolchainVersion ||
          profile.runtimeVersion !== sourceArtifact.toolchain.runtimeVersion)
    ) {
      return undefined
    }
  } else if (
    !compiledArtifact ||
    compiledArtifact.toolchain.compiler !== profile.toolchainId ||
    compiledArtifact.toolchain.compilerVersion !== profile.toolchainVersion
  ) {
    return undefined
  }
  return { hash: artifact.hash }
}

export const createDeploymentLaneIdentityResolver =
  (
    registry: Readonly<DeploymentLaneRegistry>,
  ): ((
    revision: StrategyRevision | StrategyRevisionV117,
  ) => ExecutableLaneIdentity | undefined) =>
  (revision) => {
    const profile = registry.lanes.find(
      (candidate) =>
        candidate.languageId === revision.runtime.language.id &&
        candidate.languageVersion === revision.runtime.language.version &&
        candidate.adapterId === revision.runtime.adapter.id &&
        candidate.adapterVersion === revision.runtime.adapter.version,
    )
    if (
      !profile ||
      (profile.semanticTupleId === CANDIDATE_RUNTIME_V117_SEMANTIC_TUPLE_ID &&
        !StrategyRevisionV117Schema.safeParse(revision).success) ||
      revision.metadata.providerValidation?.providerId !== profile.providerId ||
      revision.metadata.providerValidation.contractVersion.length === 0 ||
      revision.metadata.providerValidation.sourceHash !== revision.sourceHash ||
      revision.metadata.providerValidation.sourceBytes !==
        revision.sourceBytes ||
      revision.metadata.providerValidation.proof.length === 0 ||
      revision.runtime.abiVersion !== profile.semanticTuple.runtimeAbi ||
      revision.engineCompatibility.spec !== profile.semanticTuple.rules ||
      revision.engineCompatibility.engine !== profile.semanticTuple.engine
    ) {
      return undefined
    }
    const artifact = artifactBytesAndHash(revision, profile)
    if (!artifact) return undefined
    const revisionArtifact =
      profile.artifactKind === "source"
        ? revision.metadata.sourceArtifact
        : revision.metadata.compiledArtifact
    if (
      revisionArtifact === undefined ||
      revision.metadata.providerValidation.artifactHash !==
        revisionArtifact?.hash ||
      revision.metadata.providerValidation.artifactBytes !==
        revisionArtifact.bytes
    ) {
      return undefined
    }
    return Object.freeze({
      providerId: profile.providerId,
      languageId: profile.languageId,
      runtimeId: profile.runtimeId,
      runtimeVersion: profile.runtimeVersion,
      toolchainId: profile.toolchainId,
      toolchainVersion: profile.toolchainVersion,
      adapterId: profile.adapterId,
      adapterVersion: profile.adapterVersion,
      policyId: profile.policyId,
      policyVersion: profile.policyVersion,
      corpusId: profile.corpusId,
      corpusVersion: profile.corpusVersion,
      artifactId: `${profile.artifactIdPrefix}${revision.id}`,
      artifactSha256: artifact.hash,
      implementationId: profile.implementationId,
      buildId: profile.buildId,
      semanticTupleId: profile.semanticTupleId,
      semanticTuple: Object.freeze({ ...profile.semanticTuple }),
    })
  }

export const createSuccessorRuntimeIdentityTemplateResolver = (
  registry: Readonly<DeploymentLaneRegistry>,
): ((
  revision: StrategyRevision | StrategyRevisionV117,
) => DeploymentLaneProfile["successorRuntimeIdentityTemplate"] | undefined) => {
  const resolveLane = createDeploymentLaneIdentityResolver(registry)
  return (revision) => {
    if (resolveLane(revision) === undefined) return undefined
    return registry.lanes.find(
      (candidate) =>
        candidate.languageId === revision.runtime.language.id &&
        candidate.languageVersion === revision.runtime.language.version &&
        candidate.adapterId === revision.runtime.adapter.id &&
        candidate.adapterVersion === revision.runtime.adapter.version,
    )?.successorRuntimeIdentityTemplate
  }
}
